dawn/src/dawn/Compiler/DawnCompiler.cpp

```diff
@@ -170,184 +209,203 @@ std::unique_ptr<OptimizerContext> DawnCompiler::runOptimizer(std::shared_ptr<SIR
+                           stencilInstantiationMap) {
 
   // Generate code
-  std::unique_ptr<codegen::CodeGen> CG;
-
   if(options_->Backend == "gt" || options_->Backend == "gridtools") {
-    CG = std::make_unique<codegen::gt::GTCodeGen>(optimizer->getStencilInstantiationMap(),
-                                                  *diagnostics_, options_->UseParallelEP,
-                                                  options_->MaxHaloPoints);
+    auto CG = std::make_unique<codegen::gt::GTCodeGen>(
+        stencilInstantiationMap, *diagnostics_, options_->UseParallelEP, options_->MaxHaloPoints);
+    return CG->generateCode();
   } else if(options_->Backend == "c++-naive") {
-    CG = std::make_unique<codegen::cxxnaive::CXXNaiveCodeGen>(
-        optimizer->getStencilInstantiationMap(), *diagnostics_, options_->MaxHaloPoints);
+    auto CG = std::make_unique<codegen::cxxnaive::CXXNaiveCodeGen>(
+        stencilInstantiationMap, *diagnostics_, options_->MaxHaloPoints);
+    return CG->generateCode();
   } else if(options_->Backend == "c++-naive-ico") {
-    CG = std::make_unique<codegen::cxxnaiveico::CXXNaiveIcoCodeGen>(
-        optimizer->getStencilInstantiationMap(), *diagnostics_, options_->MaxHaloPoints);
+    auto CG = std::make_unique<codegen::cxxnaiveico::CXXNaiveIcoCodeGen>(
+        stencilInstantiationMap, *diagnostics_, options_->MaxHaloPoints);
+    return CG->generateCode();
   } else if(options_->Backend == "cuda") {
     const Array3i domain_size{options_->domain_size_i, options_->domain_size_j,
                               options_->domain_size_k};
-    CG = std::make_unique<codegen::cuda::CudaCodeGen>(
-        optimizer->getStencilInstantiationMap(), *diagnostics_, options_->MaxHaloPoints,
-        options_->nsms, options_->maxBlocksPerSM, domain_size);
+    auto CG = std::make_unique<codegen::cuda::CudaCodeGen>(stencilInstantiationMap, *diagnostics_,
+                                                           options_->MaxHaloPoints, options_->nsms,
+                                                           options_->maxBlocksPerSM, domain_size);
+    return CG->generateCode();
   } else if(options_->Backend == "c++-opt") {
     dawn_unreachable("GTClangOptCXX not supported yet");
   } else {
     diagnostics_->report(buildDiag("-backend", options_->Backend,
                                    "backend options must be : " +
                                        dawn::RangeToString(", ", "", "")(std::vector<std::string>{
                                            "gridtools", "c++-naive", "c++-opt", "c++-naive-ico"})));
-    return nullptr;
+    throw std::runtime_error("An error occurred.");
+  }
+}
+
+std::unique_ptr<codegen::TranslationUnit>
+DawnCompiler::compile(const std::shared_ptr<SIR>& stencilIR) {
+  diagnostics_->clear();
+  diagnostics_->setFilename(stencilIR->Filename);
+
+  IIRSerializer::Format serializationKind = IIRSerializer::Format::Json;
```

Comment:
should we extract the checks of command line arguments from here and put them in a function?

Reply:
Yes, I think we should. But, I think we should do that after merging `fight_the_indirection` into here since that's going to make a ton of merge conflicts.

Reply:
:+1: